### ISSUE TITLE
`socket` reverts to uninitialised state on Windows

### ISSUE THREAD

**Issue by jonatanklosko** [erlang/otp#8853]
**Describe the bug**

After unsuccessful `socket:connect` attempts, socket operations start to return `{error,notinitialised}` (corresponds to `WSANOTINITIALISED` in WinSock), as if WinSock was no longer initialised.

**To Reproduce**

I narrowed it down to a case where we attempt two connections. Interestingly, after the first failure everything still works as expected, but after the second attempt `socket` breaks altogether.

```erlang
-module(test).

-export([test/0]).

test() ->
    file:delete("sock1"),
    file:delete("sock2"),
    {ok, Socket1} = socket:open(local, stream),
    ok = socket:bind(Socket1, #{family => local, path => <<"sock1">>}),
    {error, _} = socket:connect(Socket1, #{family => local, path => <<"none">>}),
    {ok, Socket2} = socket:open(local, stream),
    ok = socket:bind(Socket2, #{family => local, path => <<"sock2">>}),
    {error, _} = socket:connect(Socket2, #{family => local, path => <<"none">>}),
    erlang:display(socket:open(local, stream)).
    %% => {error,notinitialised}
```

(closing each socket after connect does not help)

**Expected behavior**

Subsequent `socket:open` to return a new socket.

**Affected versions**

Tested on OTP 27.

**Additional context**

In case this matters, I am running virtualized ARM Windows 11.

**Comment by jonatanklosko** [comment]
Looking at the code `WSAStartup` is called once, when the NIF is loaded, whereas `WSACleanup` is called in case of errors. The [`WSAStartup`](https://learn.microsoft.com/en-us/windows/win32/api/winsock/nf-winsock-wsastartup) documentation states:

> When it has finished using the services of the Winsock DLL, the application must call [WSACleanup](https://learn.microsoft.com/en-us/windows/desktop/api/winsock/nf-winsock-wsacleanup) to allow the Winsock DLL to free internal Winsock resources used by the application.

So perhaps `WSACleanup` ends up being called too eagerly and effectively uninitializes WinSock.

Also, the [WSANOTINITIALISED](https://learn.microsoft.com/en-us/windows/win32/winsock/windows-sockets-error-codes-2#WSANOTINITIALISED) error description says:

> [...] or [WSACleanup](https://learn.microsoft.com/en-us/windows/desktop/api/winsock/nf-winsock-wsacleanup) has been called too many times.

Note that I am not familiar with WinSock, so these are my best guesses.

**Comment by jonatanklosko** [comment]
Hey @bmk, do you think that once this is fixed, the fix could be backported in a patch release for 26 and 27?

**Comment by bmk** [comment]
That was my intention.

**Comment by bmk** [comment]
PR solving this (I hope): bmk/esock/20240924/excessive_use_of_cleanup/otp-19251

**Comment by bmk** [comment]
A patch for this has been released with 27.1.1.
A patch for 26 will come at a later date.

**Comment by jonatanklosko** [comment]
@bmk fantastic, thank you!
